src/backend/distributed/planner/merge_planner.c

```diff
@@ -0,0 +1,702 @@
+													   List *targetList,
+													   CmdType commandType);
+#endif
+
+
+/*
+ * MergeQuerySupported does check for a MERGE command in the query, if it finds
+ * one, it will verify the below criteria
+ * - Supported tables and combinations in ErrorIfMergeHasUnsupportedTables
+ * - Distributed tables requirements in ErrorIfDistTablesNotColocated
+ * - Checks target-lists and functions-in-quals in TargetlistAndFunctionsSupported
+ */
+DeferredErrorMessage *
+MergeQuerySupported(Query *originalQuery, bool multiShardQuery,
+					PlannerRestrictionContext *plannerRestrictionContext)
+{
+	/* function is void for pre-15 versions of Postgres */
+	#if PG_VERSION_NUM < PG_VERSION_15
+
+	return NULL;
+
+	#else
+
+	/* For non-MERGE commands it's a no-op */
+	if (!QueryHasMergeCommand(originalQuery))
+	{
+		return NULL;
+	}
+
+	List *rangeTableList = ExtractRangeTableEntryList(originalQuery);
+	RangeTblEntry *resultRte = ExtractResultRelationRTE(originalQuery);
+
+	/*
+	 * Fast path queries cannot have merge command, and we prevent the remaining here.
+	 * In Citus we have limited support for MERGE, it's allowed only if all
+	 * the tables(target, source or any CTE) tables are are local i.e. a
+	 * combination of Citus local and Non-Citus tables (regular Postgres tables)
+	 * or distributed tables with some restrictions, please see header of routine
+	 * ErrorIfDistTablesNotColocated for details.
+	 */
+	DeferredErrorMessage *deferredError =
+		ErrorIfMergeHasUnsupportedTables(originalQuery,
+										 rangeTableList,
+										 plannerRestrictionContext);
+	if (deferredError)
+	{
+		/* MERGE's unsupported combination, raise the exception */
+		RaiseDeferredError(deferredError, ERROR);
+	}
+
+	Oid resultRelationId = resultRte->relid;
+	deferredError =
+		MergeActionListSupported(resultRelationId,
+								 originalQuery->jointree,
+								 originalQuery->jointree->quals,
+								 originalQuery->targetList,
+								 originalQuery->commandType);
+	if (deferredError)
+	{
+		return deferredError;
+	}
+
+	/*
+	 * MERGE is a special case where we have multiple modify statements
+	 * within itself. Check each INSERT/UPDATE/DELETE individually.
+	 */
+	MergeAction *action = NULL;
+	foreach_ptr(action, originalQuery->mergeActionList)
+	{
+		Assert(originalQuery->returningList == NULL);
+		deferredError =
+			MergeActionListSupported(resultRelationId,
+									 originalQuery->jointree,
+									 action->qual,
+									 action->targetList,
+									 action->commandType);
+		if (deferredError)
+		{
+			/* MERGE's unsupported scenario, raise the exception */
+			RaiseDeferredError(deferredError, ERROR);
+		}
+	}
+
+	deferredError =
+		InsertPartitionColumnMatchesSource(originalQuery, resultRte);
+	if (deferredError)
+	{
+		/* MERGE's unsupported scenario, raise the exception */
+		RaiseDeferredError(deferredError, ERROR);
+	}
+
+	if (multiShardQuery)
+	{
+		deferredError =
+			DeferErrorIfUnsupportedSubqueryPushdown(originalQuery,
```

Comment:
DeferErrorIfUnsupportedSubqueryPushdown already does `RestrictionEquivalenceForPartitionKeys`. Do we need the same function call in `ErrorIfTablesNotColocatedAndJoinedOnDistColumn`?

Reply:
`DeferErrorIfUnsupportedSubqueryPushdown` is called  only for multi-shard queries, let me check.

Reply:
Yeah, I remember now, we can't call `DeferErrorIfUnsupportedSubqueryPushdown` for single shard queries, here is one Assert that triggers down the lane. I didn't debug further (even the current code makes this assumption about multi-shard queries).

`TRAP: FailedAssertion("ContainsRecurringRTE(rangeTableEntry, &recurType)", File: "planner/query_pushdown_planning.c", Line: 1425, PID: 98128)
`

Reply:
Ok, removed it now, and adjusted the error message.